### ISSUE TITLE
[JENKINS-39914] Declarative pipeline does not allow loops to reuse code (verbatim or parametrized same-pattern) in the Jenkinsfile script

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/pipeline-model-definition-plugin#894]
<p>I am learning the Pipeline plugin by trying to wrap a portable project build and test routine into this paradigm. I started with the Multibranch approach and Declarative DSL syntax, and found that I can not make loops there. This issue splits to several use-cases, so I'll try to untangle this furball below:</p>

<p>1) Part of the test routine is also very patternized and parallelizable - it has to run several 'make $distcheckargs' jobs for several distcheck scenarios predefined in the project Makefile. I did manage to loop over this one as an array of strings, and ultimately even managed to build a list into a variable and then parallel() it, but only by forfeiting the DDSL syntax and using a script{} step.</p>

<p>2) The overall pipeline-contents code is same for different Unixish platforms supported by that project (including rather obscure and ancient ones) - so I want the resulting Jenkinsfile to loop over my agents with respective labels (VMs and physical hosts with those OSes), and push the same pipeline commands to them. So far I could not achieve this - I can not use script{} or parallel{} or variable assignment and loops at high levels in the pipeline code; I can not (in DDSL) nest stages to separate the OSes and contain CI steps; I can not wrap the pipeline{} block into a for-loop because pipeline{} apparently must be at the top level in a file, otherwise keywords defined for DDSL become invalid markup for the common DSL and apparently I have to rewrite the prepared DDSL code into equivalent DSL...</p>

<p>I gather something of this sort is a relatively standard FOSS use-case, so it should be supported by DDSL (and if it happens to be implemented already - then it should get more visibly and explicitly documented and exemplified).</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jimklimov.png" title="jimklimov's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jimklimov">jimklimov</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-39914" target="_blank">Declarative pipeline does not allow loops to reuse code (verbatim or parametrized same-pattern) in the Jenkinsfile script</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer">abayer</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: pipeline-model-definition-plugin
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2016-12-19T20:59:57+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 6
<li><b>imported</b>: 2025-12-06
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I am learning the Pipeline plugin by trying to wrap a portable project build and test routine into this paradigm. I started with the Multibranch approach and Declarative DSL syntax, and found that I can not make loops there. This issue splits to several use-cases, so I'll try to untangle this furball below:</p>

<p>1) Part of the test routine is also very patternized and parallelizable - it has to run several 'make $distcheckargs' jobs for several distcheck scenarios predefined in the project Makefile. I did manage to loop over this one as an array of strings, and ultimately even managed to build a list into a variable and then parallel() it, but only by forfeiting the DDSL syntax and using a script{} step.</p>

<p>2) The overall pipeline-contents code is same for different Unixish platforms supported by that project (including rather obscure and ancient ones) - so I want the resulting Jenkinsfile to loop over my agents with respective labels (VMs and physical hosts with those OSes), and push the same pipeline commands to them. So far I could not achieve this - I can not use script{} or parallel{} or variable assignment and loops at high levels in the pipeline code; I can not (in DDSL) nest stages to separate the OSes and contain CI steps; I can not wrap the pipeline{} block into a for-loop because pipeline{} apparently must be at the top level in a file, otherwise keywords defined for DDSL become invalid markup for the common DSL and apparently I have to rewrite the prepared DDSL code into equivalent DSL...</p>

<p>I gather something of this sort is a relatively standard FOSS use-case, so it should be supported by DDSL (and if it happens to be implemented already - then it should get more visibly and explicitly documented and exemplified).</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/39914?focusedId=277973&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-277973">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-39914?focusedId=277973&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-277973">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Can you give me some pseudo code example so for what you'd like to be able to do in Declarative but can't currently?</p>
</pre>
</details>

</ul>
</details>

<p>Can you give me some pseudo code example so for what you'd like to be able to do in Declarative but can't currently?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jimklimov.png" title="jimklimov's avatar" /> jimklimov:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/39914?focusedId=277977&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-277977">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-39914?focusedId=277977&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-277977">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yes, just a few minutes ago I succeeded in making a job which works and launches (or rather queues) the runs I want, link posted to IRC and here :</p>

<p>FWIW, This pipeline rolls for me as both a loop over nodelabels, and inside a loop over distcheck-like Make recipes. But it is no longer DDSL, at least not for the most part (the "environment{}" block is a handy one to keep intact though <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/wink.png" height="16" width="16" align="absmiddle" alt="" border="0"/> ) - <a href="http://pastebin.com/dSdNX4rh" class="external-link" target="_blank" rel="nofollow noopener">http://pastebin.com/dSdNX4rh</a></p>
</pre>
</details>

</ul>
</details>

<p>Yes, just a few minutes ago I succeeded in making a job which works and launches (or rather queues) the runs I want, link posted to IRC and here :</p>

<p>FWIW, This pipeline rolls for me as both a loop over nodelabels, and inside a loop over distcheck-like Make recipes. But it is no longer DDSL, at least not for the most part (the "environment{}" block is a handy one to keep intact though <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/wink.png" height="16" width="16" align="absmiddle" alt="" border="0"/> ) - <a href="http://pastebin.com/dSdNX4rh" class="external-link" target="_blank" rel="nofollow noopener">http://pastebin.com/dSdNX4rh</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/39914?focusedId=277978&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-277978">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-39914?focusedId=277978&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-277978">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>That's...impressively deranged. =) Nested parallels and stages are things we've been avoiding due in part to visualization challenges in Blue Ocean, but I'll use this as further evidence that we need to figure that visualization out! I have been thinking about a <tt>matrix</tt> section at the top that would allow you to run the normally defined stages in parallel across multiple configurations of other top level steps (I.e., agent, environment, etc) but that's most likely not going to make it into 1.0. For now, you're probably best off using standard Pipeline script, but as I said, I'll keep this use case in mind going forward.</p>
</pre>
</details>

</ul>
</details>

<p>That's...impressively deranged. =) Nested parallels and stages are things we've been avoiding due in part to visualization challenges in Blue Ocean, but I'll use this as further evidence that we need to figure that visualization out! I have been thinking about a <tt>matrix</tt> section at the top that would allow you to run the normally defined stages in parallel across multiple configurations of other top level steps (I.e., agent, environment, etc) but that's most likely not going to make it into 1.0. For now, you're probably best off using standard Pipeline script, but as I said, I'll keep this use case in mind going forward.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jimklimov.png" title="jimklimov's avatar" /> jimklimov:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/39914?focusedId=277979&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-277979">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-39914?focusedId=277979&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-277979">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks for the... compliment, I guess <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/wink.png" height="16" width="16" align="absmiddle" alt="" border="0"/> And good luck cooking up another great product <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>Thanks for the... compliment, I guess <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/wink.png" height="16" width="16" align="absmiddle" alt="" border="0"/> And good luck cooking up another great product <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/0x89.png" title="0x89's avatar" /> 0x89:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/39914?focusedId=332097&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332097">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-39914?focusedId=332097&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332097">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> are you still considering this?</p>

<p>very simple use case that I would like to see</p>

<p>instead of</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">        stage(<span class="code-quote">'distribute the image'</span>) {
            parallel {
                stage(<span class="code-quote">'copy to eastus'</span>) {
                    steps {
                        sh <span class="code-quote">'bash -xe distribute-image.sh eastus'</span>
                    }
                }
                stage(<span class="code-quote">'copy to australiaeast'</span>) {
                    steps {
                        sh <span class="code-quote">'bash -xe distribute-image.sh australiaeast'</span>
                    }
                }
            }
        }
</pre>
</div></div>

<p>I want</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">        stage(<span class="code-quote">'distribute the image'</span>) {
            parallel {
                <span class="code-keyword">for</span> (location in [<span class="code-quote">"eastus"</span>, <span class="code-quote">"australiaeast"</span>]) {
                    stage(<span class="code-quote">"copy to $location"</span>) {
                        steps {
                            sh <span class="code-quote">"bash -xe distribute-image.sh $location"</span>
                        }
                    }
                }
            }
        }
</pre>
</div></div>

<p>The syntax checker complains that only "stage" calls are allowed inside parallel - but unwinding the loop, that is the only thing that happens..
So shouldn't this be possible at least in theory?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> are you still considering this?</p>

<p>very simple use case that I would like to see</p>

<p>instead of</p>


<pre>
stage(<span class="code-quote">'distribute the image'</span>) {
    parallel {
stage(<span class="code-quote">'copy to eastus'</span>) {
    steps {
sh <span class="code-quote">'bash -xe distribute-image.sh eastus'</span>
    }
}
stage(<span class="code-quote">'copy to australiaeast'</span>) {
    steps {
sh <span class="code-quote">'bash -xe distribute-image.sh australiaeast'</span>
    }
}
    }
}
</pre>

<p>I want</p>


<pre>
stage(<span class="code-quote">'distribute the image'</span>) {
    parallel {
<span class="code-keyword">for</span> (location in [<span class="code-quote">"eastus"</span>, <span class="code-quote">"australiaeast"</span>]) {
    stage(<span class="code-quote">"copy to $location"</span>) {
steps {
    sh <span class="code-quote">"bash -xe distribute-image.sh $location"</span>
}
    }
}
    }
}
</pre>

<p>The syntax checker complains that only "stage" calls are allowed inside parallel - but unwinding the loop, that is the only thing that happens..<br/>
So shouldn't this be possible at least in theory?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/39914?focusedId=332100&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332100">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-39914?focusedId=332100&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332100">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=0x89" class="user-hover" rel="0x89">0x89</a> - <a href="https://issue-redirect.jenkins.io/issue/40986" title="Matrix structure for Declarative Pipeline" class="issue-link" data-issue-key="JENKINS-40986"><del>JENKINS-40986</del></a> is going to be what you want.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=0x89" class="user-hover" rel="0x89">0x89</a> - <a href="https://issue-redirect.jenkins.io/issue/40986" title="Matrix structure for Declarative Pipeline" class="issue-link" data-issue-key="JENKINS-40986"><del>JENKINS-40986</del></a> is going to be what you want.</p>
